### ISSUE TITLE
Proposal: LoadingButton

### ISSUE THREAD

**Issue by niels9001** [microsoft/microsoft-ui-xaml#2234]
# Proposal: 
A new ProgressButton control that can show the (indeterminate) progress ring next to the ContentPresenter to indicate that a running task.

## Summary
A button control that can show a progress ring to indicate that the action that is attached to the click event is processed. The Fluent UI toolkits has this control already:(https://fluentsite.z22.web.core.windows.net/components/button/definition?showCode=false&showRtl=false&showTransparent=false&showVariables=false#states-loading)

## Rationale
The ProgressBar and ProgressRing are great to show the user that the application is doing something. Often, these tasks (that show the progress) are attached to a specific user action, such as clicking a button. Writing the XAML to include both a e.g. ProgressRing and button can be a pain and we see a lot of different implementations of this, resulting in additional inconsistency across apps.

* Making it easier for developers to show to the user that a user invoked action is running in the control they just clicked.
* Creates better UX harmonization across applications if developers adopt the same pattern.

## Scope
| Capability | Priority |
| :---------- | :------- |
| Should inherit default button behavior | Must |
| Should work with all default ButtonStyles (e.g. RevealButtonStyle or AccentButtonStyle | Must |
| A property to indicate that the (e.g. IsLoading) | Must |
| When the IsLoading property = true, the ProgressRing should be shown | Must |
| When the IsLoading property = true, the button should be in the disabled state | Must |

## Important Notes
XAML:
`<LoadingButton Content="Send message" IsLoading="True" />`

Example visual from the Fluent UI toolkit:

![Progressbutton](https://user-images.githubusercontent.com/9866362/78500163-4adacc00-7755-11ea-91f4-3731b4d5f9cd.gif)

## Open Questions
...

**Comment by mdtauk** [comment]
And can it please replace this button, so we have some consistency
![image](https://user-images.githubusercontent.com/7389110/78502068-094b2080-7757-11ea-826d-f0ceb3dbbe48.png)

**Comment by mdtauk** [comment]
Could I also suggest that there be a transition between the Loading states of this button.

![Loading Buttons](https://user-images.githubusercontent.com/7389110/78506678-43c2b680-7773-11ea-8949-d8609ec20063.gif)
Quick animated mockup.

Things I am thinking about:
- Transitioning between the Pre-loading | Loading |  Loaded states
- Text Strings/Content for each of those three states
- Handling button sizing so each string and the ProgressRing can fit
- Possibly add a value property and a determinate state for the ProgressRing (when it is added to the control)

**Comment by harvinders** [comment]
It would also be great if such a button can be added to command bar. Imagine a download button in a `CommandBar`, which when pressed shows the download progress using progress ring.

What this means is such a button should support both determinate and indeterminate mode.

**Comment by mdtauk** [comment]
@harvinders One issue with having an AppBar**Loading**Button - is how to handle the Label.  With it being an element hidden by default, and appearing when pressing the overflow button - how would this work?

**Comment by MarissaMatt** [comment]
@niels9001 I apologize for the very late response to this proposal but would the scope of this proposal be to only have a way to include a progress ring to a button control? Or would this expand to have the progress ring be added to controls where a user would need to download or upload something?

**Comment by niels9001** [comment]
> @niels9001 I apologize for the very late response to this proposal but would the scope of this proposal be to only have a way to include a progress ring to a button control? Or would this expand to have the progress ring be added to controls where a user would need to download or upload something?

This proposal is about adding a progress ring into a Button (and maybe @AppBarButton as @harvinders pointed out).

**Comment by MarissaMatt** [comment]
Thank you @niels9001! This sounds more like a potential control that would fall under button so I assigned @kikisaints to look at this more.

**Comment by mdtauk** [comment]
The content of the button should probably change between the Rest, Loading, and Loaded states - it would be useful if this content could be set in XAML, and not require code behind to change the content strings.

**Comment by chingucoding** [comment]
> The content of the button should probably change between the Rest, Loading, and Loaded states - it would be useful if this content could be set in XAML, and not require code behind to change the content strings.

Yes, totally agree with @mdtauk . I think having a RestContent, LoadingContent and LoadedContent property or something like that is the easiest way for developers.

**Comment by mdtauk** [comment]
![example loading button 1](https://camo.githubusercontent.com/f0b8dd151dc2026ccb5176083d81bd1ee3e71273/68747470733a2f2f63646e2e7261776769742e636f6d2f6d6174686965756475746f75722f72656163742d70726f67726573732d627574746f6e2f6d61737465722f6578616d706c652f64656d6f2e676966)

![example loading button 2](https://user-images.githubusercontent.com/6943919/55510393-97686480-5634-11e9-8676-af96d39306df.gif)

![example loading button 3](https://iosexample.com/content/images/2020/05/sysdefault.gif)

**Comment by heartacker** [comment]
is it hard to implement?

**Comment by castorix** [comment]
> is it hard to implement?

You can just add a ProgressRing as Button Content :

![ButtonProgressRing](https://user-images.githubusercontent.com/22345506/232683643-4b075a9c-9865-4f09-8470-b2148c5a2ebb.gif)
